wallet-new/src/Cardano/Wallet/Kernel/Wallets.hs

```diff
@@ -126,9 +127,20 @@ createHdWallet pw mnemonic spendingPassword assuranceLevel walletName = do
     -- STEP 1: Generate the 'EncryptedSecretKey' outside any acid-state
     -- transaction, to not leak it into acid-state's transaction logs.
     let (_, esk) = safeDeterministicKeyGen (BIP39.mnemonicToSeed mnemonic) spendingPassword
-    -- STEP 2: Atomically generate the wallet and the initial internal structure in
-    -- an acid-state transaction.
+
+    -- STEP 2: Insert the key into the keystore. We do this preemptively so that,
+    -- in case of asynchronous exceptions, the worst which can happen is for the
+    -- key to stay dangling into the keystore and the wallet not to be created.
+    -- Then calling 'createHdWallet' a second time in this situation would
+    -- correctly persist the wallet in the DB.
```

Comment:
Let's also point here the even more subtle issue with interleaving (i.e. scheduling) and the fact that if `applyBlock` would be called immediately after creating the wallet (without they key, i.e.) this would be very bad as the code would panic.